asserts/snapasserts/validation_sets.go

```diff
@@ -0,0 +1,244 @@
+	buf := bytes.NewBufferString("validation sets are in conflict:")
+	for _, err := range e.Snaps {
+		fmt.Fprintf(buf, "\n- %v", err)
+	}
+	return buf.String()
+}
+
+// ValidationSets can hold a combination of validation-set assertions
+// and can check for conflicts or help applying them.
+type ValidationSets struct {
+	// sets maps sequence keys to validation-set in the combination
+	sets map[string]*asserts.ValidationSet
+	// snaps maps snap-ids to snap constraints
+	snaps map[string]*snapContraints
+}
+
+const presConflict asserts.Presence = "conflict"
+
+var unspecifiedRevision = snap.R(0)
+var invalidPresRevision = snap.R(-1)
+
+type snapContraints struct {
+	name     string
+	presence asserts.Presence
+	// revisions maps revisions to pairing of ValidationSetSnap
+	// and the origination validation-set key
+	// * unspecifiedRevision is used for constraints without a
+	//   revision
+	// * invalidPresRevision is used for constraints that mark
+	//   presence as invalid
+	revisions map[snap.Revision][]*revConstraint
+}
+
+type revConstraint struct {
+	validationSetKey string
+	asserts.ValidationSetSnap
+}
+
+func (c *snapContraints) conflict() (conflicting []string, err error) {
+	if c.presence != presConflict {
+		return nil, nil
+	}
+
+	const dontCare asserts.Presence = ""
+	whichSets := func(rcs []*revConstraint, presence asserts.Presence) string {
+		which := make([]string, 0, len(rcs))
+		for _, rc := range rcs {
+			if presence != dontCare && rc.Presence != presence {
+				continue
+			}
+			conflicting = append(conflicting, rc.validationSetKey)
+			which = append(which, rc.validationSetKey)
+		}
+		if len(which) == 0 {
+			return ""
+		}
+		sort.Strings(which)
+		return fmt.Sprintf("(%s)", strings.Join(which, ","))
+	}
+
+	msg := fmt.Sprintf("cannot constraint snap %q", c.name)
+	invalid := false
+	if invalidOnes, ok := c.revisions[invalidPresRevision]; ok {
+		msg += fmt.Sprintf(" as both invalid %s and required", whichSets(invalidOnes, asserts.PresenceInvalid))
+		invalid = true
+	}
+
+	var revnos []int
+	for r := range c.revisions {
+		if r.N >= 1 {
+			revnos = append(revnos, r.N)
+		}
+	}
+	if len(revnos) == 1 {
+		msg += fmt.Sprintf(" at revision %d %s", revnos[0], whichSets(c.revisions[snap.R(revnos[0])], dontCare))
+	} else if len(revnos) > 1 {
+		sort.Ints(revnos)
+		l := make([]string, 0, len(revnos))
+		for _, rev := range revnos {
+			l = append(l, fmt.Sprintf("%d %s", rev, whichSets(c.revisions[snap.R(rev)], dontCare)))
+		}
+		msg += fmt.Sprintf(" at different revisions %s", strings.Join(l, ", "))
+	}
+
+	if unspecifiedOnes, ok := c.revisions[unspecifiedRevision]; ok {
+		which := whichSets(unspecifiedOnes, asserts.PresenceRequired)
+		if which != "" {
+			if len(revnos) != 0 {
+				msg += " or"
+			}
+			if invalid {
+				msg += fmt.Sprintf(" at any revision %s", which)
+			} else {
+				msg += fmt.Sprintf(" required at any revision %s", which)
+			}
+		}
+	}
+	return conflicting, fmt.Errorf(msg)
```

Comment:
Nitpick, while safe for other reasons, `fmt.Errorf` should not be used this way, as `msg` may by accident contain formatting escape codes.